plugins/outputs/azuremonitor/azuremonitor.go

```diff
@@ -0,0 +1,445 @@
+}
+
+type azureMonitorSeries struct {
+	DimensionValues []string `json:"dimValues"`
+	Min             float64  `json:"min"`
+	Max             float64  `json:"max"`
+	Sum             float64  `json:"sum"`
+	Count           int64    `json:"count"`
+}
+
+// Write writes metrics to the remote endpoint
+func (a *AzureMonitor) Write(metrics []telegraf.Metric) error {
+	azmetrics := make(map[uint64]*azureMonitorMetric, len(metrics))
+	for _, m := range metrics {
+		id := hashIDWithTagKeysOnly(m)
+		if azm, ok := azmetrics[id]; !ok {
+			azmetrics[id] = translate(m)
+		} else {
+			azmetrics[id].Data.BaseData.Series = append(
+				azm.Data.BaseData.Series,
+				translate(m).Data.BaseData.Series...,
+			)
+		}
+	}
+
+	var body []byte
+	for _, m := range azmetrics {
+		// Azure Monitor accepts new batches of points in new-line delimited
+		// JSON, following RFC 4288 (see https://github.com/ndjson/ndjson-spec).
+		jsonBytes, err := json.Marshal(&m)
+		if err != nil {
+			return err
+		}
+		body = append(body, jsonBytes...)
+		body = append(body, '\n')
+	}
+
+	if err := a.validateCredentials(); err != nil {
+		return fmt.Errorf("E! Unable to fetch authentication credentials: %v", err)
+	}
+
+	req, err := http.NewRequest("POST", a.url, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+a.msiToken.AccessToken)
+	req.Header.Set("Content-Type", "application/x-ndjson")
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 300 || resp.StatusCode < 200 {
+		reply, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			reply = nil
+		}
+		return fmt.Errorf("E! Get Error. %d HTTP response: %s response body: %s",
+			resp.StatusCode, resp.Status, reply)
+	}
+
+	return nil
+}
+
+func hashIDWithTagKeysOnly(m telegraf.Metric) uint64 {
+	h := fnv.New64a()
+	h.Write([]byte(m.Name()))
+	h.Write([]byte("\n"))
+	for _, tag := range m.TagList() {
+		h.Write([]byte(tag.Key))
+		h.Write([]byte("\n"))
+	}
+	b := make([]byte, binary.MaxVarintLen64)
+	n := binary.PutUvarint(b, uint64(m.Time().UnixNano()))
+	h.Write(b[:n])
+	h.Write([]byte("\n"))
+	return h.Sum64()
+}
+
+func translate(m telegraf.Metric) *azureMonitorMetric {
+	var dimensionNames []string
+	var dimensionValues []string
+	for i, tag := range m.TagList() {
+		// Azure custom metrics service supports up to 10 dimensions
+		if i > 10 {
+			log.Printf("W! [outputs.azuremonitor] metric [%s] exceeds 10 dimensions", m.Name())
+			continue
+		}
+		dimensionNames = append(dimensionNames, tag.Key)
+		dimensionValues = append(dimensionValues, tag.Value)
+	}
+
+	min, _ := m.GetField("min")
+	max, _ := m.GetField("max")
+	sum, _ := m.GetField("sum")
+	count, _ := m.GetField("count")
+	return &azureMonitorMetric{
+		Time: m.Time(),
+		Data: &azureMonitorData{
+			BaseData: &azureMonitorBaseData{
+				Metric:         m.Name(),
+				Namespace:      "default",
```

Comment:
Any reason we choose to set the metric namespace to be "default"?  This might be good for users to override via config but as a default, it might be better to go with a value which has "Telegraf" in the namespace.